File: Editorial/99_Recover_Binary_Search_Tree/CSharp/2_Iterative_Inorder_Traversal.cs
public class Solution {
    public void Swap(ref TreeNode a, ref TreeNode b) {
        int tmp = a.val;
        a.val = b.val;
        b.val = tmp;
    }

    public void RecoverTree(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode x = null, y = null, pred = null;
        while (stack.Count != 0 || root != null) {
            while (root != null) {
                stack.Push(root);
                root = root.left;
            }

            root = stack.Pop();
            if (pred != null && root.val < pred.val) {
                y = root;
                if (x == null)
                    x = pred;
                else
                    break;
            }

            pred = root;
            root = root.right;
        }

        Swap(ref x, ref y);
    }
}

File: Editorial/0000-0100/0099_Recover_Binary_Search_Tree/CPP/2_Iterative_Inorder_Traversal.cpp
class Solution {
public:
    void recoverTree(TreeNode* root) {
        stack<TreeNode*> stack;
        TreeNode *x = NULL, *y = NULL, *pred = NULL;
        while (!stack.empty() || root != NULL) {
            while (root != NULL) {
                stack.push(root);
                root = root->left;
            }
            root = stack.top();
            stack.pop();
            if (pred != NULL && root->val < pred->val) {
                y = root;
                if (x == NULL)
                    x = pred;
                else
                    break;
            }
            pred = root;
            root = root->right;
        }
        swap(x->val, y->val);
    }
};
